openmmtools/cache.py

```diff
@@ -0,0 +1,352 @@
+#!/usr/bin/env python
+
+# =============================================================================
+# MODULE DOCSTRING
+# =============================================================================
+
+"""
+Provide cache classes to handle creation of OpenMM Context objects.
+
+"""
+
+
+# =============================================================================
+# GLOBAL IMPORTS
+# =============================================================================
+
+import copy
+import collections
+
+from simtk import openmm
+
+from openmmtools import utils
+
+
+# =============================================================================
+# GENERAL LRU CACHE
+# =============================================================================
+
+class LRUCache(object):
+    """A simple LRU cache.
+
+    It can be configured to have a maximum number of elements (capacity)
+    and an element expiration (time_to_live) measured in number of accesses
+    to the cache. Both read and write operations count as an access, but
+    only successful reads (i.e. those not raising KeyError) increment the
+    counter.
+
+    Parameters
+    ----------
+    capacity : int, optional
+        Maximum number of elements in the cache. When set to None, the
+        cache has infinite capacity (default is None).
+    time_to_live : int, optional
```

Comment:
Does it make sense to later be able to add an alternative `time_to_live` measured in actual wall clock time? I guess we could interpret this as number of accesses if it is an `int` or later interpret it as wall clock time if a `simtk.unit.seconds` compatible quantity is specified.